File: ovp8xx/cpp/scc/scc_calibration.cpp
/*
 * Copyright 2025-present ifm electronic, gmbh
 * SPDX-License-Identifier: Apache-2.0
 */
/* This example shows how to calibrate a camera
 * using the SCC application.
 * The calibration is triggered by sending the
 * "calibrate" commande in JSON with the Set function.
 * The result is received in the O3R_RESULT_JSON
 * buffer, through the previously registered callback.
 * The generated values can be written to the device
 * using the "writeToDevice" command, in JSON.
 */

#include <fstream>
#include <ifm3d/common/json.hpp>
#include <ifm3d/device/o3r.h>
#include <ifm3d/fg.h>
#include <iostream>
#include <memory>

using namespace ifm3d::literals;
using namespace std::chrono_literals;
using ifm3d::json;

// Global variable to track calibration status
bool CALIBRATED = false;

// Function to send a command to the SCC application
void send_command(std::shared_ptr<ifm3d::O3R> o3r,
                  const std::string &app_instance, const std::string &command) {
  try {
    json command_config = {
        {"applications",
         {{"instances",
           {{app_instance, {{"configuration", {{"command", command}}}}}}}}}};
    o3r->Set(command_config);
    std::cout << "Command '" << command << "' sent successfully" << std::endl;
  } catch (const std::exception &e) {
    std::cerr << "Error while sending command '" << command << "': " << e.what()
              << std::endl;
    throw;
  }
}

// Callback function to process calibration results
void calibration_callback(ifm3d::Frame::Ptr frame) {
  if (frame->HasBuffer(ifm3d::buffer_id::O3R_RESULT_JSON) && !CALIBRATED) {
    auto result_json_buffer =
        frame->GetBuffer(ifm3d::buffer_id::O3R_RESULT_JSON);
    std::string result_json_string = std::string(
        result_json_buffer.ptr<char>(0),
        strnlen(result_json_buffer.ptr<char>(0), result_json_buffer.size()));
    ifm3d::json result_json = ifm3d::json::parse(result_json_string);

    // Check the calibrationState value
    int calibration_state = result_json.value("calibrationState", 0);
    if (calibration_state == 1) {
      CALIBRATED = true;
      std::cout << "Calibration successful!" << std::endl;
      std::cout << "Calibration Results:" << std::endl
                << result_json.dump(4) << std::endl; // Pretty-print JSON
    } else {
      std::cout << "Calibration failed: calibrationState = "
                << calibration_state << std::endl;
    }
  }
}

int main() {
  // IP address of the VPU
  std::string IP = "192.168.0.69"; // Default IP
  // Path to the configuration file, assuming
  // that the example is built in scc/build
  std::string config_path = "../config/scc_calibration_port2.json";

  // Create the O3R device object
  auto o3r = std::make_shared<ifm3d::O3R>(IP);

  // Load configuration
  std::ifstream file(config_path);
  if (!file) {
    throw std::runtime_error("Failed to open config file: " + config_path);
  }
  json config = json::parse(file);
  file.close();

  // Retrieve the application instance
  auto instances = config["applications"]["instances"];
  if (instances.empty()) {
    throw std::runtime_error(
        "No applications found in the configuration file.");
  }
  std::string app_instance = instances.begin().key();
  std::cout << "Using application instance: " << app_instance << std::endl;

  // Ensure a clean start, only reseting the instance we are using
  try {
    auto current_instances = o3r->Get({"/applications/instances"});
    if (current_instances["applications"]["instances"].contains(app_instance)) {
      o3r->Reset("/applications/instances/" + app_instance);
      std::cout << "Reset application instance: " << app_instance << std::endl;
    } else {
      std::cout << "Application instance '" << app_instance
                << "' not found, skipping reset." << std::endl;
    }
  } catch (const std::exception &e) {
    std::cerr << "Error while resetting application: " << e.what() << std::endl;
  }

  // Apply configuration
  try {
    o3r->Set(config);
    std::cout << "Configuration applied successfully" << std::endl;
  } catch (const std::exception &e) {
    std::cerr << "Error while applying configuration: " << e.what()
              << std::endl;
    return 1;
  }

  // Set application to RUN state
  try {
    json run_state = {{"applications",
                       {{"instances", {{app_instance, {{"state", "RUN"}}}}}}}};
    o3r->Set(run_state);
    std::cout << "Application set to RUN state" << std::endl;
  } catch (const std::exception &e) {
    std::cerr << "Error while setting application to RUN state: " << e.what()
              << std::endl;
    return 1;
  }

  // Clear the buffer before calibration
  send_command(o3r, app_instance, "clearBuffer");

  // Setup framegrabber
  auto fg = std::make_shared<ifm3d::FrameGrabber>(
      o3r, o3r->Port(app_instance).pcic_port);
  fg->Start({ifm3d::buffer_id::O3R_RESULT_JSON});
  fg->OnNewFrame(&calibration_callback);

  // Calibrate the camera
  try {
    send_command(o3r, app_instance, "calibrate");
  } catch (const std::exception &e) {
    std::cerr << "Error during calibration: " << e.what() << std::endl;
    return 1;
  }

  if (CALIBRATED) {
    std::cout << "Calibration successful, writing to device..." << std::endl;
    // Write calibration to device
    try {
      send_command(o3r, app_instance, "writeToDevice");
    } catch (const std::exception &e) {
      std::cerr << "Error while writing calibration to device: " << e.what()
                << std::endl;
      return 1;
    }
  } else {
    std::cout << "Calibration failed, please check the camera setup."
              << std::endl;
  }

  // Stop the framegrabber
  fg->Stop();

  return 0;
}

File: ovp8xx/python/scc/scc_calibration.py
# -*- coding: utf-8 -*-
#############################################
# Copyright 2025-present ifm electronic, gmbh
# SPDX-License-Identifier: Apache-2.0
#############################################
# This script is calibrating a camera using
# the SCC embedded application.
# The calibration routine is triggered by sending the
# "calibrate" command to the SCC application with the
# set command, in JSON.
# The result is received in the O3R_RESULT_JSON buffer,
# through the registered callback.
# The "writeToDevice" command can be sent to set
# the generated calibration values to the port.

import json
import pathlib

import numpy as np
from ifm3dpy.device import O3R, Error
from ifm3dpy.framegrabber import FrameGrabber, buffer_id

# Global variable to track calibration status
CALIBRATED = False


def load_config(config_path):
    """Load configuration from a JSON file."""
    try:
        with open(config_path, "r") as f:
            return json.load(f)
    except OSError as err:
        print("Error while reading configuration file")
        raise err


def calibration_callback(frame):
    """Callback to process received calibration results."""
    global CALIBRATED
    if frame.has_buffer(buffer_id.O3R_RESULT_JSON) and not CALIBRATED:
        json_chunk = frame.get_buffer(buffer_id.O3R_RESULT_JSON)
        json_array = np.frombuffer(json_chunk[0], dtype=np.uint8).tobytes()
        parsed_json_array = json.loads(json_array.decode())
        # Check the calibrationState value
        calibration_state = parsed_json_array.get("calibrationState", 0)
        if calibration_state == 1:
            CALIBRATED = True
            print("Calibration successful!")
            print(json.dumps(parsed_json_array, indent=4))  # Pretty-print JSON


def send_command(o3r, app_instance, command):
    """Send a command to the application."""
    try:
        o3r.set(
            {
                "applications": {
                    "instances": {app_instance: {"configuration": {"command": command}}}
                }
            }
        )
        print(f"Command '{command}' sent successfully")
    except Error as err:
        print(f"Error while sending command '{command}'")
        raise err


def main(ip, config_file):
    global CALIBRATED
    # Initialize O3R device
    o3r = O3R(ip)

    # Load configuration
    config_path = pathlib.Path(__file__).parent / config_file
    config_snippet = load_config(config_path)
    app_instance = next(iter(config_snippet["applications"]["instances"]))
    print(f"Using application instance: {app_instance}")

    # Ensure a clean start, only reseting the instance we are using
    try:
        current_instances = o3r.get(["/applications/instances"])["applications"]
        if current_instances is not None:
            instances = current_instances["instances"].keys()
            if app_instance in instances:
                o3r.reset("/applications/instances/" + app_instance)
                print(f"Reset application instance: {app_instance}")
            else:
                print(
                    f"Application instance '{app_instance}' not found, skipping reset."
                )
        else:
            print("No application instances found, skipping reset.")
    except Error as e:
        print(f"Reset failed: {e}")

    # Apply configuration
    try:
        o3r.set(config_snippet)
        print("Configuration applied successfully")
    except Error as err:
        print("Error while applying configuration")
        raise err

    # Set application to RUN state
    try:
        o3r.set(
            {
                "applications": {
                    "instances": {app_instance: {"class": "scc", "state": "RUN"}}
                }
            }
        )
        print("Application set to RUN state")
    except Error as err:
        print("Error while setting application to RUN state")
        raise err

    # Clear the buffer before calibration to not have any old data
    send_command(o3r, app_instance, "clearBuffer")

    # Setup framegrabber
    fg = FrameGrabber(o3r, o3r.port(app_instance).pcic_port)
    fg.start([buffer_id.O3R_RESULT_JSON])
    fg.on_new_frame(calibration_callback)

    send_command(o3r, app_instance, "calibrate")

    if CALIBRATED:
        print("Calibration successful, writing to device...")
        # Write calibration to device
        send_command(o3r, app_instance, "writeToDevice")
    else:
        print("Calibration failed, please check the camera setup.")

    # stop framegrabber
    fg.stop()


if __name__ == "__main__":
    ip = "192.168.0.69"
    config_file = "config/scc_calibration_port2_with_translations.json"
    main(ip, config_file)
